pkg/boot/kexec/memory_linux.go

```diff
@@ -365,13 +365,18 @@ func (m *Memory) LoadElfSegments(r io.ReaderAt) error {
 			continue
 		}
 
-		d := make([]byte, p.Filesz)
-		n, err := r.ReadAt(d, int64(p.Off))
-		if err != nil {
-			return err
-		}
-		if n < len(d) {
-			return fmt.Errorf("not all data of the segment was read")
+		var d []byte
+		if p.Filesz == 0 {
+			d = make([]byte, p.Memsz)
+		} else {
+			d = make([]byte, p.Filesz)
```

Comment:
Are you sure about this? When memsz > filesz, I thought it's supposed to fill the rest with zeros, which is a behavior not retained in this PR

Reply:
oh, nevermind. i got it. had to read the code below. notice how the segment is made with Paddr and Memsz, not Filesz.

You don't need to have `d` be a slice of 0s -- kexec pads segments that have bufsz < memsz with 0s. So a kexec_segment as below fills the purpose, and `d` can be `nil`, which should be converted to `buf: 0, bufsz: 0`.

{
buf: 0
bufsz: 0
mem: 0x1000
memsz: 0x1000
}

the kernel knows how to interpret that in kexec - fill the area with 0s.

Did you encounter a specific bug with this?

Reply:
Perhaps it'd be a good idea to quote the man-page in the code for future readers:

```
bufsz bytes are copied from the source buffer to the target kernel
          buffer.  If bufsz is less than memsz, then the excess bytes in the
          kernel buffer are zeroed out.
```

http://man7.org/linux/man-pages/man2/kexec_load.2.html

Reply:
Thanks for the detailed analysis, I didn't know the details of kexec load.

I had a bug with my build of the multiboot test kernel (build with buildroot so a different gcc and libc). The resulting elf is triggering it with a section that was misinterpreted by the parser.
I initially got the error `error loading ELF segments: bad offset 8192` but as the size is 0, the offset is irrelevant.